src/device/mod.rs

```diff
@@ -0,0 +1,323 @@
+// Copyright 2018 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE-BSD-3-Clause file.
+//
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+//! A module that offers building blocks for virtio devices.
+
+mod mmio;
+mod virtio_config;
+
+use vm_memory::GuestAddressSpace;
+
+use std::result;
+use std::sync::atomic::AtomicU8;
+use std::sync::Arc;
+
+use crate::Queue;
+
+pub use mmio::VirtioMmioDevice;
+pub use virtio_config::{VirtioConfig, VirtioDeviceActions, VirtioDeviceType};
+
+// TODO: Bring this (and other feature definitions) to the vm-virtio crate proper.
+// Using a local const temporarily until then.
+const VIRTIO_F_RING_EVENT_IDX: u64 = 29;
+
+/// When the driver initializes the device, it lets the device know about the completed stages
+/// using the Device Status field.
+///
+/// These following consts are defined in the order in which the bits would typically be set by
+/// the driver. `RESET` -> `ACKNOWLEDGE` -> `DRIVER` and so on. This module is a 1:1 mapping for
+/// the Device Status field in the virtio 1.1 specification, section 2.1 (except for the `RESET`
+/// value, which is not explicitly named there as such). The status flag descriptions (except
+/// `RESET`) are taken from the standard.
+pub mod status {
+    /// The initial status of the device.
+    pub const RESET: u8 = 0;
+    /// Indicates that the guest OS has found the device and recognized it as a valid
+    /// virtio device.
+    pub const ACKNOWLEDGE: u8 = 1;
+    /// Indicates that the guest OS knows how to drive the device.
+    pub const DRIVER: u8 = 2;
+    /// Indicates that something went wrong in the guest, and it has given up on the device.
+    /// This could be an internal error, or the driver didn’t like the device for some reason,
+    /// or even a fatal error during device operation.
+    pub const FAILED: u8 = 128;
```

Comment:
Could we sort these consts by value?

Reply:
They were initially written down in the order they are presented in the standard (i.e. in section 2.1), but we can def reorder them if it improves clarity & readability.